File: dogfight/src/entities/runway.ts
import { FacingDirection, Team } from "../constants";
import { TypedEntity, EntityType } from "../TypedEntity";
import { Cache, CacheEntry } from "../network/cache";
import { SolidEntity } from "./SolidEntity";
import { spriteSheet } from "../../../client/src/render/textures";
import { Rectangle } from "../physics/rectangle";
import { SimplePlane } from "pixi.js";

const RESERVE_TAKEOFF_LANDING_DELAY = 1000;
const RESERVE_LANDING_TAKEOFF_DELAY = 2000;
const RESERVE_LANDING_LANDING_DELAY = 1000;
const RESERVE_TAKEOFF_TAKEOFF_DELAY = 1000;
const TAKEOFF = 1;
const LANDING = 1;
const HEALTH_MAX = 1530;
const HEALTH_TIMER_MAX = 50;

export class Runway extends SolidEntity {
  public type = EntityType.Runway;

  public x: number;
  public y: number;
  public direction: FacingDirection;
  public team: Team;
  public health: number;

  public lastReserve: number;
  public reserveTimer: number;
  public healthTimer: number;


  public image = [spriteSheet.textures["runway.gif"], spriteSheet.textures["runway2.gif"]];
  public imageWidth = [this.image[0].width, this.image[1].width];
  public imageHeight = [this.image[0].height, this.image[1].height];

  public constructor(id: number, cache: Cache, team: number, x: number, y: number, direction: number) {
    super(id, team);
    this.setData(cache, {
      x: x,
      y: y,
      direction: direction, //FacingDirection.Right,
      team: team, //Team.Centrals,
      health: 1530
    });
  }

  public getImageWidth(paramInt: number): number {
    return this.imageWidth[paramInt];
  }

  public getCollisionBounds(): Rectangle {
    return new Rectangle(this.x, this.y, this.imageWidth[1 - this.direction], this.imageHeight[1 - this.direction]);
  }

  public getImage() {
    return this.image[1 - this.direction];
  }

  public getLandableWidth(): number {
    return this.imageWidth[1 - this.direction] - 65;
  }

  public getLandableX(): number {
    if (this.direction == 1) {
      return 65 + this.x;
    }
    return this.x;
  }
  public getLandableY(): number {
    return 23 + this.y;
  }
  public getStartX(): number {
    if (this.direction == 1) {
      return 15 + this.x;
    }
    return this.x + 230;
  }
  public getStartY(): number {
    return this.getLandableY();
  }
  public getDirection(): number {
    return this.direction;
  }

  /**
   * TODO synchronized?!?!?
   * @param paramInt LANDING or TAKEOFF request
   */
  public reserveFor(paramInt: number): boolean {
    let l1 = Date.now();
    let l2 = 0;
    switch (paramInt) {
      case 2:
        if (this.lastReserve == 2) {
          l2 = 1000;
        } else {
          l2 = 1000;
        }
        break;
      case 1:
        if (this.lastReserve == 2) {
          l2 = 2000;
        } else {
          l2 = 1000;
        }
        break;
    }
    if (this.reserveTimer + l2 > l1) {
      return false;
    }
    if (this.health <= 0) {
      return false;
    }
    this.reserveTimer = l1;
    this.lastReserve = paramInt;
    return true;
  }

  public run(): void {
    if ((this.health > 0) && (this.health < 1530)) {
      this.healthTimer = ((this.healthTimer + 1) % 50);
      if (this.healthTimer == 0) {
        this.health += 1;
        if (this.health * 255 % 1530 == 0) {
          //setChanged(true);
        }
      }
    }
  }

  public hit(paramSolidEntity: SolidEntity): void {
    if (this.health <= 0) {
      return;
    }
    switch (paramSolidEntity.getType()) {
      case 17:
        this.health -= 30;
        break;
      case EntityType.Bullet:
        this.health -= (int)(4.0D * ((Bullet)paramSolidEntity).getDamageFactor());
        break;
      case 27:
        this.health -= 17;
        break;
      default:
        return;
    }
    if (this.health <= 0) {
      this.health = 0;
      this.destroyed(paramSolidEntity.getTeam());
    }
    setChanged(true);
  }

  public getState(): CacheEntry {
    return {
      type: this.type,
      x: this.x,
      y: this.y,
      direction: this.direction,
      team: this.team,
      health: this.health
    };
  }
}

File: dogfight/src/entities/runway.rs
use dogfight_macros::Networked;
use image::RgbaImage;

use crate::{
    collision::{BoundingBox, SolidEntity},
    images::{get_image, RUNWAY, RUNWAY2},
    math::get_client_percentage,
    network::{property::*, EntityProperties, NetworkedEntity},
};

use super::{
    entity::Entity,
    types::{EntityType, Facing, Team},
};

#[derive(Debug)]
pub enum RunwayReservation {
    Takeoff = 1,
    Landing = 2,
}

// Yes, these are named like in the original, somewhat confusing
const RESERVE_TAKEOFF_LANDING_DELAY: i32 = 1000;
const RESERVE_LANDING_TAKEOFF_DELAY: i32 = 2000;
const RESERVE_LANDING_LANDING_DELAY: i32 = 1000;
const RESERVE_TAKEOFF_TAKEOFF_DELAY: i32 = 1000;

#[derive(Networked)]
pub struct Runway {
    team: Property<Team>,
    facing: Property<Facing>,
    client_x: Property<i16>,
    client_y: Property<i16>,
    client_health: Property<u8>,

    total_health: i32,
    health_timer: i32,

    last_reserve: Option<RunwayReservation>,
    reserve_timer: i32,

    image_runway: RgbaImage,
    image_runway2: RgbaImage,
}

impl Runway {
    pub fn new(team: Team, facing: Facing, x: i16, y: i16) -> Self {
        let mut runway = Runway {
            team: Property::new(team),
            facing: Property::new(facing),
            client_x: Property::new(x),
            client_y: Property::new(y),
            client_health: Property::new(0),
            total_health: 0,
            health_timer: 0,
            image_runway: get_image(RUNWAY),
            image_runway2: get_image(RUNWAY2),
            last_reserve: None,
            reserve_timer: 10_000, // setting this to a high value to allow the player to immediately take off on game load
        };

        runway.set_health(runway.get_max_health());

        runway
    }

    pub fn tick(&mut self) -> () {
        self.reserve_timer += 10;

        if self.total_health > 0 && self.total_health < self.get_max_health() {
            self.health_timer = (self.health_timer + 1) % 50;

            if self.health_timer == 0 {
                self.set_health(self.total_health + 1);
            }
        }
    }

    pub fn subtract_health(&mut self, amount: i32) -> () {
        let new_health = self.total_health - amount;
        self.set_health(new_health);
    }

    pub fn set_health(&mut self, health_amount: i32) -> () {
        self.total_health = health_amount;

        if self.total_health < 0 {
            self.total_health = 0;
        }

        self.client_health.set(get_client_percentage(
            self.total_health,
            self.get_max_health(),
        ));
    }

    fn get_max_health(&self) -> i32 {
        1530
    }

    pub fn get_team(&self) -> &Team {
        self.team.get()
    }

    pub fn get_client_x(&self) -> i16 {
        *self.client_x.get()
    }

    pub fn get_client_y(&self) -> i16 {
        *self.client_y.get()
    }

    pub fn get_landable_x(&self) -> i16 {
        let x = self.client_x.get();
        match self.facing.get() {
            Facing::Left => *x,
            Facing::Right => x + 65,
        }
    }

    pub fn get_landable_y(&self) -> i16 {
        self.client_y.get() + 23
    }

    pub fn get_start_x(&self) -> i16 {
        // original offsets are 15 for right facing, and 230 for left facing,
        // but these don't seem quite right, at least the left facing one.
        // probably just a difference between how planes X,Y are calculated
        // in the original vs this one.
        match self.facing.get() {
            Facing::Right => self.client_x.get() + 20,
            Facing::Left => self.client_x.get() + 255,
        }
    }

    pub fn get_facing(&self) -> Facing {
        *self.facing.get()
    }

    pub fn get_start_y(&self) -> i16 {
        self.get_landable_y()
    }

    pub fn get_landable_width(&self) -> i16 {
        self.get_image().width() as i16 - 65
    }

    fn get_image(&self) -> &RgbaImage {
        match self.facing.get() {
            Facing::Right => &self.image_runway,
            Facing::Left => &self.image_runway2,
        }
    }

    // Lentokonepeli Air traffic controller logic :)
    pub(crate) fn reserve_for(&mut self, reservation: RunwayReservation) -> bool {
        let delay = match reservation {
            RunwayReservation::Landing => {
                if let Some(RunwayReservation::Landing) = self.last_reserve {
                    RESERVE_TAKEOFF_LANDING_DELAY
                } else {
                    RESERVE_TAKEOFF_TAKEOFF_DELAY
                }
            }
            RunwayReservation::Takeoff => {
                if let Some(RunwayReservation::Takeoff) = self.last_reserve {
                    RESERVE_LANDING_TAKEOFF_DELAY
                } else {
                    RESERVE_LANDING_LANDING_DELAY
                }
            }
        };

        /*
        log(format!(
            "reserve_timer: {}, delay: {}, ",
            self.reserve_timer, delay
        ));
        */

        if self.reserve_timer < delay {
            return false;
        }

        if self.is_alive() == false {
            return false;
        }

        /*
        log(format!(
            "delay: {delay}, reservation: {:?}, last_reserve: {:?} timer: {}",
            reservation, self.last_reserve, self.reserve_timer
        ));
        */

        self.reserve_timer = 0;
        self.last_reserve = Some(reservation);
        // Clear for landing/takeoff :)
        return true;
    }
}

impl Entity for Runway {
    fn get_type(&self) -> EntityType {
        EntityType::Runway
    }
}

impl Runway {
    pub fn get_landing_bounds(&self) -> BoundingBox {
        BoundingBox {
            x: self.get_landable_x(),
            y: self.get_start_y(),
            width: self.get_landable_width(),
            height: self.get_start_y() - self.get_landable_y(),
        }
    }
}

impl SolidEntity for Runway {
    fn get_collision_bounds(&self) -> BoundingBox {
        let img = self.get_image();
        BoundingBox {
            x: *self.client_x.get(),
            y: *self.client_y.get(),
            width: img.width() as i16,
            height: img.height() as i16,
        }
    }

    fn get_collision_image(&self) -> Option<&RgbaImage> {
        Some(self.get_image())
    }

    fn do_rotate_collision_image(&mut self) -> () {}

    fn is_alive(&self) -> bool {
        self.total_health > 0
    }
}
